### PR TITLE
workflows: add nightly linux tests

### PR DESCRIPTION
**Pull request by iMichka**
Test a simple simple brew installation and "brew install hello" step for various distributions.

This makes sure everything is working well.
We use our installer as we want to test everything end to end.

The idea is also to be able to tell users how to install curl/git/tar on older linux versions when needed.

In the future, these images might be uploaded to our docker registry, and I plan to use them
to test migrations (from gcc5 to gcc7, and potentially for building binutils on ubuntu 18.04).

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
I had to go through some hoops to get this working on older linuxes.

I have still 2 issues:
- I would like to use make the test.sh script fail on error. Using `set -e` works, but makes the whole script fail on the following unrelated error: `bash: cannot set terminal process group (-1): Inappropriate ioctl for device / bash: no job control in this shell`
- The centos 7 test fails: glibc DOES NOT get automatically installed, thus it is missing. There is also a postinstall failure in gcc@5 that I need to check, which happens when you try to install glibc.

**Comment by Bo98** [comment]
> * would like to use make the test.sh script fail on error. Using `set -e` works, but makes the whole script fail on the following unrelated error: `bash: cannot set terminal process group (-1): Inappropriate ioctl for device / bash: no job control in this shell`

This happens becuase you're running it as interactive (`bash -i`) even though CI is not an interactive environment. Is that intentional?

**Comment by Bo98** [comment]
> * The centos 7 test fails: glibc DOES NOT get automatically installed, thus it is missing.

This is known because of the incorrect conditional in gcc@5. But we should consider moving that dependency to `brew` and not be tied to installing gcc@5.

**Comment by danielnachun** [comment]
> > * The centos 7 test fails: glibc DOES NOT get automatically installed, thus it is missing.
> 
> This is known because of the incorrect conditional in gcc@5. But we should consider moving that dependency to `brew` and not be tied to installing gcc@5.

This is a good idea and fits with some previous discussion about treating `glibc` as a vendored dependency like `portable-ruby` rather than a regular formula.

I think we may also have a problem where `gcc@5` can end up being installed **before** `glibc` and then tries to compile it instead of the host toolchain.  Disaster ensues because our spec adds `RPATH` to the `ld.so` which completely breaks it.  Brewed `gcc` using our custom spec should never be used to build `glibc`.  It could be worth adding logic to make sure that the default spec is used for `gcc` if the user tries to use it to build `glibc`.

**Comment by danielnachun** [comment]
I think the other thing we should think about is how to better handle users getting access to newer `curl` and `git` on systems without root access.  Right now I've been using Miniconda (ironically) to "bootstrap" a newer `curl` and `git` so I can install brewed `curl` and `git`.  This works well enough but I feel like we should have a better way of handling this, though I'm not sure what this would be yet.  This was not an issue until we became more strict about `curl` and `git` versions and I wonder if those restrictions could be relaxed only during bootstrapping as we have done for Debian Wheezy, but I don't know how feasible that is.

**Comment by danielnachun** [comment]
A last thought for now: is it reasonable for us to test a CentOS 7 image in a non-default prefix without root?  I know we don't officially support this, but just making sure the bootstrapping process works to the point where `brew` is usable would be nice.  Binary patching of prefixes is now merged, and I'm working on the design ideas for the next step of bottling with longer prefixes so I'm hoping we'll have better support for non-default prefixes soon.

File: Library/Homebrew/nightly-linux/centos7/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,54 @@
+FROM centos:7
```
Comment (MikeMcQuaid) [inline Library/Homebrew/nightly-linux/centos7/Dockerfile:1]: What about putting this in `.github/workflows/nightly-linux/centos7/Dockerfile` instead?

File: Library/Homebrew/nightly-linux/centos7/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+# hadolint ignore=DL3008
```
Comment (MikeMcQuaid) [inline Library/Homebrew/nightly-linux/centos7/Dockerfile:3]: Can we get comments whenever we are ignoring linters. Also, is `hadolint` setup to validate these?

File: Library/Homebrew/nightly-linux/centos7/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    && cd / \
+    && rm -rf /curl-7.82.0
+
+RUN curl -sL http://mirrors.edge.kernel.org/pub/software/scm/git/git-2.28.0.tar.gz | tar xz \
+    && cd /git-2.28.0 \
+    && make configure \
+    && ./configure --prefix=/usr/local --with-curl=/usr/local \
+    && make install NO_TCLTK=1 \
+    && cd / \
+    && rm -rf /git-2.28.0
```
Comment (MikeMcQuaid) [inline Library/Homebrew/nightly-linux/centos7/Dockerfile:35]: I don't think we should be building these from source every night, we should come up with a better solution here for Linux systems.

File: Library/Homebrew/nightly-linux/centos7/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+RUN /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+
+RUN chown -R linuxbrew /home/linuxbrew/.linuxbrew
+
+RUN echo "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /home/linuxbrew/.bashrc
+RUN echo "export HOMEBREW_GIT_PATH=/usr/local/bin/git" >> /home/linuxbrew/.bashrc
+RUN echo "export HOMEBREW_CURL_PATH=/usr/local/bin/curl" >> /home/linuxbrew/.bashrc
+RUN echo "export HOMEBREW_FORCE_BREWED_CA_CERTIFICATES=1" >> /home/linuxbrew/.bashrc
+# This is necessary so that HOMEBREW_GIT_PATH and HOMEBREW_CURL_PATH are picked up by brew:
+RUN echo "export HOMEBREW_DEVELOPER=1" >> /home/linuxbrew/.bashrc
```
Comment (MikeMcQuaid) [inline Library/Homebrew/nightly-linux/centos7/Dockerfile:54]: Add newlines here.

File: Library/Homebrew/nightly-linux/centos7/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+RUN su - linuxbrew -c 'mkdir ~/.linuxbrew'
+
+RUN /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+
+RUN chown -R linuxbrew /home/linuxbrew/.linuxbrew
+
+RUN echo "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /home/linuxbrew/.bashrc
+RUN echo "export HOMEBREW_GIT_PATH=/usr/local/bin/git" >> /home/linuxbrew/.bashrc
+RUN echo "export HOMEBREW_CURL_PATH=/usr/local/bin/curl" >> /home/linuxbrew/.bashrc
+RUN echo "export HOMEBREW_FORCE_BREWED_CA_CERTIFICATES=1" >> /home/linuxbrew/.bashrc
```
Comment (MikeMcQuaid) [inline Library/Homebrew/nightly-linux/centos7/Dockerfile:52]: How will users know they need to set these?

File: Library/Homebrew/nightly-linux/test.sh [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
```
Comment (MikeMcQuaid) [inline Library/Homebrew/nightly-linux/test.sh:2]: `set -e`

File: Library/Homebrew/nightly-linux/test.sh [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+brew install hello
+brew test hello
+brew uninstall hello
+brew install -s hello
+brew test hello
```
Comment (MikeMcQuaid) [inline Library/Homebrew/nightly-linux/test.sh:7]: Feels like `brew test-bot` may be a better fit here?

**Review by MikeMcQuaid — COMMENTED** [review]
I'm not opposed to this approach but I have a few concerns:
- CentOS 7 and Debian 7,8 both require a bunch of compiling from source and variables to be set for them to work. I think we should decide if we definitely want to support them and, if we do, provide a better solution e.g. recommended backports, fixing it so `brew install curl git` works on those OSs or, worst case, `portable-curl` and `portable-git` or something rather than expect users to figure this out
- who fixes these builds when they fail? This unfortunately seems to be perfectly setup for a situation where these builds are created but never actually get verified/fixed
- if we're only running these on a schedule: how do we identify what broke them?

**Comment by iMichka** [comment]
> This happens becuase you're running it as interactive (bash -i) even though CI is not an interactive environment. Is that intentional?

Yes. I did not want to `ADD` or `COPY` the `test.sh` script to the image, and run the script directly with `docker run`. This is the only way I found. Is there an alternative?

**Comment by iMichka** [comment]
>> The centos 7 test fails: glibc DOES NOT get automatically installed, thus it is missing.
>This is known because of the incorrect conditional in gcc@5. But we should consider moving that dependency to brew and not be tied to installing gcc@5.

I see. I have a first fix for that: https://github.com/Homebrew/homebrew-core/pull/100787

**Comment by Bo98** [comment]
> > This happens becuase you're running it as interactive (bash -i) even though CI is not an interactive environment. Is that intentional?
> 
> Yes. I did not want to `ADD` or `COPY` the `test.sh` script to the image, and run the script directly with `docker run`. This is the only way I found. Is there an alternative?

Not entirely sure how the `-i` flag would have any effect in that but I'll have a look.

**Comment by iMichka** [comment]
> Not entirely sure how the -i flag would have any effect in that but I'll have a look.

I remember now. It is because I wanted the env variables from the .bashrc file to be loaded. If you don't use an interactive shell the env variables are not set and brew is not found.

**Comment by Bo98** [comment]
Can you set `BASH_ENV` instead?

**Comment by iMichka** [comment]
The `.bashrc` file has this at the top on ubuntu:
```
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
```

The env variables added at the end of the file are not read.
This is why I had to add the interactive shell flag to `docker run`.

I found a workaround based on `BASH_ENV` and a second `.linuxbrew-bashrc` file for non-interactive shells. It's a little bit more verbose, we might be able to simplify things.
But this allowed to remove the interactive shell and use `set -e` in the test script.

File: Library/Homebrew/nightly-linux/test.sh (hunk 6 above)
Comment (iMichka) [reply Library/Homebrew/nightly-linux/test.sh:3]: Done

**Comment by iMichka** [comment]
> I think we may also have a problem where gcc@5 can end up being installed before glibc and then tries to compile it instead of the host toolchain. Disaster ensues because our spec adds RPATH to the ld.so which completely breaks it. Brewed gcc using our custom spec should never be used to build glibc. It could be worth adding logic to make sure that the default spec is used for gcc if the user tries to use it to build glibc.

This is out of scope of this PR

> I think the other thing we should think about is how to better handle users getting access to newer curl and git on systems without root access.

This too

> A last thought for now: is it reasonable for us to test a CentOS 7 image in a non-default prefix without root? 

This too

We can discuss this somewhere else, I don't want to pollute this PR too much and avoid getting side-tracked. Let's focus on shipping a first working version of these tests.

**Comment by iMichka** [comment]
> I'm not opposed to this approach but I have a few concerns:

> * CentOS 7 and Debian 7,8 both require a bunch of compiling from source and variables to be set for them to work. I think we should decide if we definitely want to support them and, if we do, provide a better solution e.g. recommended backports, fixing it so brew install curl git works on those OSs or, worst case, portable-curl and portable-git or something rather than expect users to figure this out

Agreed. We need to automatise all these steps / ship pre-complied binaries for git/curl. I won't be able to do all of this and will need help, and it will take some time. Having these tests is a first step in that direction, and gives us the opportunity to iron things out. The env variables should also be handled automatically by brew.

My goal is to work on the glibc/gcc-7/ubuntu18 change first (the tests in the PR will help for that), and then I can help improving the situation for legacy users (this can also be done in parallel in anyone wants to help out).

> * who fixes these builds when they fail? This unfortunately seems to be perfectly setup for a situation where these builds are created but never actually get verified/fixed
> * if we're only running these on a schedule: how do we identify what broke them?

I thought about that but have no clear idea for this right now. I did not want to run these tests on each pull request in brew, as it's maybe a bit heavy. Maybe sending an email / slack message on failure would be enough? Or maybe run them only when we merge a PR to master.

I prefer having an imperfect test setup for a first shot to having no tests at all.

**Comment by danielnachun** [comment]
Here is my best attempt at summarizing the options I think we have for getting `git`/`curl` to users on older systems without root access:

**Build from source**
**Pros:** Requires no changes to `brew` or new formulae.
**Cons:** Slow and prone to failure across diverse systems.

**Use a Miniconda environment**
**Pros:** Provides binaries that will work on almost all systems.
**Cons:** Amounts to offloading the bootstrapping problem to another package manager.

**Use `HOMEBREW_INSTALL_FROM_API` to avoid `git` and a Ruby gem to download packages instead of `curl`**
**Pros:** No dependency on `git` and `curl` during bootstrapping.
**Cons:** `HOMEBREW_INSTALL_FROM_API` is still experimental and there may not be a suitable Ruby gem that can substitute for `curl`.

**Lower or eliminate version requirements for `git` and `curl` until brewed versions are installed**
**Pros:** Would try to use whatever `git` or `curl` is available.
**Cons:** Doesn't help if the system is missing `git` or `curl`, could have security or compatibility issues.

**Provide static portable `curl` and `git` binaries build against `glibc@2.13`, possibly pinned at our minimum version**
**Pros:** Provides binaries that will work on almost all systems.
**Cons:** Requires making new formulae and maintaining them.

I currently use the second option because it works reliably, but that's probably not suitable for us, and I'd love to see the the third option eventually be how it works, but I think we're not there yet.  To me the last option is the best compromise for now, and the maintenance burden can be offset through automation.

**Comment by MikeMcQuaid** [comment]
> Agreed. We need to automatise all these steps / ship pre-complied binaries for git/curl. I won't be able to do all of this and will need help, and it will take some time. Having these tests is a first step in that direction, and gives us the opportunity to iron things out. The env variables should also be handled automatically by brew.

I'm not convinced these tests are the best first step here, though, or at least not on the platforms that require manually compiling.

> > * who fixes these builds when they fail? This unfortunately seems to be perfectly setup for a situation where these builds are created but never actually get verified/fixed
> > * if we're only running these on a schedule: how do we identify what broke them?
> 
> I thought about that but have no clear idea for this right now. I did not want to run these tests on each pull request in brew, as it's maybe a bit heavy. Maybe sending an email / slack message on failure would be enough? Or maybe run them only when we merge a PR to master.

That still doesn't really attach ownership here. I'd really like you and at least once other person in @Homebrew/linux to take responsibility for owning these builds otherwise they are just going to break all the time.

> I prefer having an imperfect test setup for a first shot to having no tests at all.

I disagree, I'm afraid. Tests suites that are too broad or too flaky or unowned are worse than no tests because they produce noise that will just be ignored.

Do we really need 11 builds here? How often do newer versions break and older versions not? Could we not just test the oldest version of each distro we support? That, combined with some better caching, feels like it could be run on every PR.

> I currently use the second option because it works reliably, but that's probably not suitable for us

I think it's fine, personally, and I'd rather see us take that approach in these CI jobs.

**Comment by danielnachun** [comment]
> > I currently use the second option because it works reliably, but that's probably not suitable for us
> 
> I think it's fine, personally, and I'd rather see us take that approach in these CI jobs.

@iMichka if you're okay with bootstrapping with Miniconda, I can try to push the changes I think are needed for this approach once I've got it worked out.  I'll probably start with a bash script and then convert the steps over to a Dockerfile.

> Do we really need 11 builds here? How often do newer versions break and older versions not? Could we not just test the oldest version of each distro we support? That, combined with some better caching, feels like it could be run on every PR.

I agree that we're probably testing too many versions, and could probably just cover the oldest versions of CentOS, Debian and Ubuntu.  I think it would be better to test other branches of the Linux "family tree" with supported Docker images: 
OpenSUSE: https://hub.docker.com/r/opensuse/leap
Arch LInux: https://hub.docker.com/_/archlinux
Gentoo: https://hub.docker.com/r/gentoo/stage3

These other distros don't have LTS versions so they'd let us test on newer platforms and maximize our coverage of different distros.

**Comment by MikeMcQuaid** [comment]
> I agree that we're probably testing too many versions, and could probably just cover the oldest versions of CentOS, Debian and Ubuntu.

I guess even here I'm wondering: when do these actually break? Have we broken them with Homebrew/brew changes recently?

This seems like a lot of overhead for something I'm not convinced is actually solving a problem we have. I'd really rather see the effort instead go into ensuring that e.g. the Ruby/Curl/Git experience is easier on these platforms.

**Comment by iMichka** [comment]
> if you're okay with bootstrapping with Miniconda,

I'd rather not depend on another package manager to bootstrap our own package manager.

> This seems like a lot of overhead for something I'm not convinced is actually solving a problem we have. I'd really rather see the effort instead go into ensuring that e.g. the Ruby/Curl/Git experience is easier on these platforms.

Agreed.

I think we should focus on getting curl and git shipped for these platforms.

I propose to
- create a new repo under `Homebrew/homebrew-bootstrap`
- move the portable ruby creation / upload code to that repo and archive the portable ruby repo
- add code to build a static curl and git
- install the newly created binaries when necessary

Thoughts?

**Comment by iMichka** [comment]
> I guess even here I'm wondering: when do these actually break? Have we broken them with Homebrew/brew changes recently?

We did not break anything recently. But the whole `binutils` changes in homebrew-core and the ubuntu 18.04 migration make me nervous. I would like to have a minimal safety net before shipping things to our users.

**Comment by MikeMcQuaid** [comment]
> * add code to build a static curl and git

What about downloading these existing formulae over HTTP and building them there. Alternatively, I wonder if we could make `curl` or `git` `cellar :any`?

> But the whole `binutils` changes in homebrew-core

This wouldn't catch homebrew-core breakages though, right?



> and the ubuntu 18.04 migration make me nervous

Is there a tracking issue for the ongoing work here?

**Comment by danielnachun** [comment]
> > * add code to build a static curl and git
> 
> What about downloading these existing formulae over HTTP and building them there. Alternatively, I wonder if we could make `curl` or `git` `cellar :any`?

Downloading over HTTP would be a great way to make this work, though we'll have to make sure that works for the dependencies of `curl` and `git` as well (including `glibc` on older Linux).

`curl` is already `cellar :any`, but `git` is not.   The bigger issue is that `curl` and `git` also have the following non-relocatable dependencies (the tree is big so this may not be 100% right):
```
openldap
util-linux
openssl@1.1
krb5
libidn2
ncurses
gettext
xz
```
I am not sure we can get all of these to be `cellar :any` without prefix patching.  

> > But the whole `binutils` changes in homebrew-core
> 
> This wouldn't catch homebrew-core breakages though, right?
> 
> > and the ubuntu 18.04 migration make me nervous
> 
> Is there a tracking issue for the ongoing work here?

I will open one when I have a moment as I do think we need to delineate exactly what needs to be done for the migration to happen.

**Comment by MikeMcQuaid** [comment]
> Downloading over HTTP would be a great way to make this work, though we'll have to make sure that works for the dependencies of `curl` and `git` as well (including `glibc` on older Linux).

Yes 😬 

> I am not sure we can get all of these to be `cellar :any` without prefix patching.

We could consider removing some of these dependencies potentially?

---

Given the above, though, I am leaning a lot more towards at least needing `portable-curl` (which could be used to `brew install git`) whereas I was historically opposed.


> * create a new repo under `Homebrew/homebrew-bootstrap`
> * move the portable ruby creation / upload code to that repo and archive the portable ruby repo

This repo used to be called `homebrew-portable` and maybe even contained `curl` and `git` and we can consider just renaming it back and adding them there, I think.

I think we should avoid building/shipping `portable-curl` and, if we decide we need it, `portable-git` on any macOS versions, though.

> I will open one when I have a moment as I do think we need to delineate exactly what needs to be done for the migration to happen.

Perfect thanks!

**Comment by danielnachun** [comment]
> We could consider removing some of these dependencies potentially?
>
> Given the above, though, I am leaning a lot more towards at least needing `portable-curl` (which could be used to `brew install git`) whereas I was historically opposed.
> 
> > * create a new repo under `Homebrew/homebrew-bootstrap`
> > * move the portable ruby creation / upload code to that repo and archive the portable ruby repo
> 
> This repo used to be called `homebrew-portable` and maybe even contained `curl` and `git` and we can consider just renaming it back and adding them there, I think.
> 
> I think we should avoid building/shipping `portable-curl` and, if we decide we need it, `portable-git` on any macOS versions, though.

I also think that's the easiest way to get this to work well in the short term.  I think a stripped down `portable-curl` and `portable-git` with the bare minimum set of dependencies needed to provide the functionality needed for `brew` would hugely simplify the bootstrapping process.  Although we would statically link both packages, we would still need to build them with `glibc@2.13` so they could run on older systems, because `glibc` cannot be statically linked.

One question for this is whether we should pin the versions of `portable-curl` and `portable-git` to the minimum versions we require (like we do with `portable-ruby`) or keep them on the sam version as the regular `curl` and `git` formulae.

It's probably also worth mentioning that we also require minimum versions of `make`, `gawk`, and `sed` to build `glibc` that could be nice to provide as portable formulae.  And we also internally use `grep`, `tar`, `gzip` and `file` for various operations in `brew`.  I don't think we need go that far down the rabbit hole yet but it may be worth fully enumerating all the executables that we rely on during the bootstrapping process if we ever wanted to a ship a fully self contained sharball installer.

This is highly speculative, but it may eventually be possible to eliminate the reliance on these entirely if we can successfully create a relocatable `glibc` bottle (it sort of works already, but has some minor issues) and a small program written in Go, which is statically linked and does not use `glibc`, to do prefix patching as I recently implemented in `brew`.  We would distribute those two items along with `portable-ruby`, the regular `curl` and `git` bottles and their dependencies, the bottles for the other programs `brew` needs, and a statically linked bottle for `patchelf` in a sharball.  After extracting the sharball, the Go program would be used to patch prefix references for `glibc` and the other non-relocatable bottles and `patchelf` would be used to set the loader path to point to brewed `glibc` and the RPATHs to `HOMEBREW_PREFIX/lib`.   

This essentially amounts to reimplementing the bottle relocation functionality of `brew` outside of `brew` but with statically linked binaries that don't rely on the host `glibc`.  Obviously there's a lot work that would need to be done make this possible but it would allow us to bootstrap without building from source into any prefix on any Linux system with kernel 2.6 or newer regardless of `glibc` version, so I figure it's worth putting out there as a future direction.  For now we need to focus on `portable-curl` and `portable-git` (and sorry for the wall of text, it's hard to explain all this concisely!).

**Comment by MikeMcQuaid** [comment]
> I also think that's the easiest way to get this to work well in the short term. I think a stripped down `portable-curl` and `portable-git` with the bare minimum set of dependencies needed to provide the functionality needed for `brew` would hugely simplify the bootstrapping process. Although we would statically link both packages, we would still need to build them with `glibc@2.13` so they could run on older systems, because `glibc` cannot be statically linked.

I still think it's worth separating the requirements somewhat. A real `git` can be installed easily from a bottle or from source with `portable-curl`, perhaps even as part of the installer. I think we should try to see if we can do with just `portable-curl` rather than adding them both potentially unnecessarily.

> One question for this is whether we should pin the versions of `portable-curl` and `portable-git` to the minimum versions we require (like we do with `portable-ruby`) or keep them on the sam version as the regular `curl` and `git` formulae.

I feel like ideally they'd be kept in sync with the regular formulae somehow.

> It's probably also worth mentioning that we also require minimum versions of `make`, `gawk`, and `sed` to build `glibc` that could be nice to provide as portable formulae. And we also internally use `grep`, `tar`, `gzip` and `file` for various operations in `brew`. I don't think we need go that far down the rabbit hole yet but it may be worth fully enumerating all the executables that we rely on during the bootstrapping process if we ever wanted to a ship a fully self contained sharball installer.

Please no 🙉 😂 

Let's keep this scope really tight and focused on the minimal number of things that users currently need. I haven't heard anyone complaining about an outdated/missing e.g. `grep` maybe ever. It's not worth spending our time building things that aren't solving actual problems for actual, multiple users.

> This essentially amounts to reimplementing the bottle relocation functionality of `brew` outside of `brew` but with statically linked binaries that don't rely on the host `glibc`.

This is a really clever idea. I agree with you it's a bit too soon for now but could be nice to have as something that would let us use the existing bottles more widely.



> For now we need to focus on `portable-curl` and `portable-git` (and sorry for the wall of text, it's hard to explain all this concisely!).

👍🏻 (except maybe don't even need `portable-git` yet).

**Comment by danielnachun** [comment]
> I still think it's worth separating the requirements somewhat. A real `git` can be installed easily from a bottle or from source with `portable-curl`, perhaps even as part of the installer. I think we should try to see if we can do with just `portable-curl` rather than adding them both potentially unnecessarily.

Right now I guess the issue is that `brew` won't run at all without a new enough `git`, so we'd either have to change this behavior or install `git` outside of `brew`, as we already do with `portable-ruby` and plan to do with `portable-curl`.  Our `git` bottle would have to be built against `glibc@2.13` to be usable without brewed `glibc` being installed, so I think it might be easier to build this from source.

> I feel like ideally they'd be kept in sync with the regular formulae somehow.

If we're using the same livecheck URL for both `curl` and `portable-curl`, I think this would happen automatically?  The only thing I don't know is how things might differ if `portable-curl` is in a different repo than `curl`.

> Please no 🙉 😂
> 
> Let's keep this scope really tight and focused on the minimal number of things that users currently need. I haven't heard anyone complaining about an outdated/missing e.g. `grep` maybe ever. It's not worth spending our time building things that aren't solving actual problems for actual, multiple users.

Yeah I haven't seen any complaints yet so I don't consider this worth pursuing right now.

> This is a really clever idea. I agree with you it's a bit too soon for now but could be nice to have as something that would let us use the existing bottles more widely.

This is definitely something I want to look into more once we've completed the 18.04 migration and fully implemented building with longer prefixes.

**Comment by MikeMcQuaid** [comment]
> Right now I guess the issue is that `brew` won't run at all without a new enough `git`, so we'd either have to change this behavior or install `git` outside of `brew`, as we already do with `portable-ruby` and plan to do with `portable-curl`.

`brew update` used to be able to bootstrap a Homebrew/brew and Homebrew/homebrew-core created from tarballs so it could be that we do the same thing here again.



> If we're using the same livecheck URL for both `curl` and `portable-curl`, I think this would happen automatically? The only thing I don't know is how things might differ if `portable-curl` is in a different repo than `curl`.

I think @Bo98 is working on improving this sort of thing for homebrew-portable-ruby and adding a CI job there for the oldest macOS version we support/can virtualise.



> Yeah I haven't seen any complaints yet so I don't consider this worth pursuing right now.

👍🏻 



> This is definitely something I want to look into more once we've completed the 18.04 migration and fully implemented building with longer prefixes.

👍🏻

**Comment by Bo98** [comment]
> I think @Bo98 is working on improving this sort of thing for homebrew-portable-ruby and adding a CI job there for the oldest macOS version we support/can virtualise.

Yes, I'm going to investigate the feasibility of Yosemite CI on Thursday.

**Comment by MikeMcQuaid** [comment]
> Yes, I'm going to investigate the feasibility of Yosemite CI on Thursday.

@Bo98 Awesome! FYI: if Yosemite doesn't work here: if there's another reasonable older version we can virtualise, I think that'd be a good enough reason to bump up the minimum macOS version we build for.

**Comment by Bo98** [comment]
I gnerally agree with the discussion here but on this note:

> Lower or eliminate version requirements for `git` and `curl` until brewed versions are installed

I feel the current minimum is poorly documented on Linux and needs clarifying why exactly it's 7.41.0, except on Debian 7.

This doesn't take away from the case for `portable-curl`. I feel it's worth investigating too given this might be the version we check before installing `portable-curl`.

**Comment by iMichka** [comment]
Thanks for all the comments, but can I get a review on the pull request? I feel the discussion is drifting in the wrong direction.

I wanted to submit a first bunch of tests, even if not perfect. They document the current way of installing homebrew. I feel there is also a lot of bikeshedding here, so please try to focus :)
I know you are all unhappy with the lack of automation, the numerous manual steps to install git/curl, and no clear documentation / checks in brew to make this all work seamlessly for our users, but this is the current state and I think the tests help document this, and are a good starting point to work on improving the Homebrew installation.

I think the main question right now is to know where to put these tests, as @MikeMcQuaid pointed out. It's tricky because we are testing a combination of 3 things: the installer script, Homebrew/brew, and Homebrew/homebrew-core. I found not better way than nightly tests.

**Comment by MikeMcQuaid** [comment]
> Thanks for all the comments, but can I get a review on the pull request? I feel the discussion is drifting in the wrong direction.

@iMichka That's fair, to be more explicit on what I'd want to see before this is merged:
- agreed ownership on who is going to fix these builds when they fail
- agreed timescale if these builds fail for e.g. 1 week that they can be disabled so they don't keep emailing people every day
- a smaller/minimum number of distress we support (i.e. oldest distro of each variety)
- not building ruby/curl/git from source each time by using e.g. Anaconda, portable versions or just actions caching
- at least two links to examples of things that would have been caught by these tests if we'd have had them in place

I don't think any of this means "portable-curl needs done first" and I realise the above comments seem harsh but: I genuinely believe merging this as-is would cause more harm than good, unfortunately.

**Pull request closed without merging**